tests/src/test/java/docs/javadsl/ConsumerExampleTest.java

```diff
@@ -0,0 +1,584 @@
+                    .via(business()))
+        .runWith(Sink.ignore(), materializer);
+
+    // #restartSource
+    assertDone(produceString(topic, 10, partition0()));
+    // #restartSource
+    control.get().shutdown();
+    // #restartSource
+  }
+
+  // #withRebalanceListenerActor
+  static class RebalanceListener extends AbstractLoggingActor {
+
+    @Override
+    public Receive createReceive() {
+      return receiveBuilder()
+          .match(
+              TopicPartitionsAssigned.class,
+              assigned -> {
+                log().info("Assigned: {}", assigned);
+              })
+          .match(
+              TopicPartitionsRevoked.class,
+              revoked -> {
+                log().info("Revoked: {}", revoked);
+              })
+          .build();
+    }
+  }
+
+  // #withRebalanceListenerActor
+
+  @Test
+  public void withRebalanceListener() throws Exception {
+    ConsumerSettings<String, String> consumerSettings =
+        consumerDefaults().withGroupId(createGroupId(1));
+    String topic = createTopic(1, 1, 1);
+    int messageCount = 10;
+    // #withRebalanceListenerActor
+    ActorRef rebalanceListener = system.actorOf(Props.create(RebalanceListener.class));
+
+    Subscription subscription =
+        Subscriptions.topics(topic)
+            // additionally, pass the actor reference:
+            .withRebalanceListener(rebalanceListener);
+
+    // use the subscription as usual:
+    Consumer.DrainingControl<List<ConsumerRecord<String, String>>> control =
+        Consumer.plainSource(consumerSettings, subscription)
+            // #withRebalanceListenerActor
+            .take(messageCount)
+            // #withRebalanceListenerActor
+            .toMat(Sink.seq(), Keep.both())
+            .mapMaterializedValue(Consumer::createDrainingControl)
+            .run(materializer);
+    // #withRebalanceListenerActor
+    assertDone(produceString(topic, messageCount, partition0()));
+    assertDone(control.isShutdown());
+    assertEquals(messageCount, resultOf(control.drainAndShutdown(ec)).size());
+  }
+
+  @Test
+  public void consumerMetrics() throws Exception {
+    ConsumerSettings<String, String> consumerSettings =
+        consumerDefaults().withGroupId(createGroupId(1));
+    String topic = createTopic(1, 1, 1);
+    // #consumerMetrics
+    // run the stream to obtain the materialized Control value
+    Consumer.DrainingControl<Done> control =
+        Consumer.plainSource(
+                consumerSettings, Subscriptions.assignment(new TopicPartition(topic, 0)))
+            .via(business())
+            .toMat(Sink.ignore(), Keep.both())
+            .mapMaterializedValue(Consumer::createDrainingControl)
+            .run(materializer);
+
+    // #consumerMetrics
+    sleepMillis(
+        100,
+        "to let the control establish itself (fails with `java.lang.IllegalStateException: not yet initialized: only setHandler is allowed in GraphStageLogic constructor)` otherwise");
```

Comment:
We should fix this or at least track this. Created https://github.com/akka/alpakka-kafka/issues/668